### PR TITLE
Improve parity with ExecutionContext and add combinators

### PR DESCRIPTION
**Pull request by RaasAhsan**
This PR addresses several things:
1. Recover parity of `ExecutionContext`s in the codebase. The `RTS` required users to override an `ExecutorService` if they want to configure their own pool. The `shift` combinator requires an `ExecutionContext`.
2. Having the `RTS` rely on an `ExecutionContext` means that we can't use `awaitTermination` or `shutdown` on the pools, so my proposal in #305 is included here as well.
3. Introduce the combinator `IO.yield` (yield is a valid Scala keyword, so backticks are actually needed. Any better names?), which is an alias for `IO.sleep(0.seconds)` to yield to another thread
4. Introduce the combinator `IO#on` that makes it easier to shift back and forth between the `RTS` pool and other pools.

Most or all of this will probably be superseded by future work on `RTS`, but I think its a good starting point for what it could look like. Let me know if its any good.

EDIT: Maybe this should have been split into two PRs. let me know

### CODE REVIEW

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 1]
```diff
@@ -913,6 +916,12 @@ object IO {
       )
     )
 
+  /**
+   * Yields execution to another thread.
+   */
+  final def `yield`: IO[Nothing, Unit] =
+    IO.sleep(0.seconds)
```
Comment (kaishh) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: This will unnecessary fork into scheduled pool. Why not 
```scala
IO.async { cb: Callback[Nothing, Unit] => cb(ExitResult.Completed(()) }
```
?

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 2]
```diff
@@ -308,6 +308,9 @@ sealed abstract class IO[+E, +A] { self =>
   final def ensuring(finalizer: IO[Nothing, Unit]): IO[E, A] =
     new IO.Ensuring(self, finalizer)
 
+  final def on(ec: ExecutionContext): IO[E, A] =
```
Comment (kaishh) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: It might make sense to also add `forkOn`

```scala
final def forkOn(ec: ExecutionContext): IO[Nothing, Fiber[E, A]] =
  on(ec).fork
```

File: core/jvm/src/main/scala/scalaz/zio/RTS.scala [hunk 3]
```diff
@@ -58,7 +52,12 @@ trait RTS {
   /**
    * The main thread pool used for executing fibers.
    */
-  val threadPool = newDefaultThreadPool()
+  val executionContext: ExecutionContext = ExecutionContext.fromExecutorService(newDefaultThreadPool())
```
Comment (kaishh) [inline core/jvm/src/main/scala/scalaz/zio/RTS.scala]: Does this mean automatic support for `scala.concurrent.blocking` contexts?

**Comment by kaishh** [comment]
> Introduce the combinator IO.yield (yield is a valid Scala keyword, so backticks are actually needed. Any better names?)

`cats-effect` uses `IO.shift` and it's legal to have both `IO.shift` and `IO.shift(ec: ExecutionContext)` at the same time. I think it's a good name (and also familiar for existing users of cats-effect)

File: core/shared/src/main/scala/scalaz/zio/IO.scala (hunk 1 above)
Comment (RaasAhsan) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: The JVM `RTS` checks for zero duration sleeps and will submit to the regular pool instead: https://github.com/scalaz/scalaz-zio/blob/master/core/jvm/src/main/scala/scalaz/zio/RTS.scala#L88 . 

I'm not sure what this means for JS.

File: core/shared/src/main/scala/scalaz/zio/IO.scala (hunk 2 above)
Comment (RaasAhsan) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: I like this one. The implementation perhaps feels a bit weird because it might not be necessary to shift back to the default pool just to product the result.

Also, what do we want the order of operations to be? Shift, fork, then shift back? Or fork, shift, and shift back?

Comment (RaasAhsan) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: Now that I think about it, shift and then fork doesn't make any sense because fork submits the task on the default pool.

Comment (kaishh) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: > Users of this function would need to be aware that inserting an IO.shift would shift execution to the RTS pool.

Yes, but we might want to change that in the future, I.e. make fiber remember its execution context to have better compatibility with `cats-effect` and libraries that want the user to supply an explicit ec - currently at least fs2, http4s, doobie.

Comment (kaishh) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: > The implementation perhaps feels a bit weird because it might not be necessary to shift back to the default pool just to product the result.

Ah, yes, that's redundant. It should probably be just:

```scala
(IO.shift(ec) *> self).fork
```

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 4]
```diff
@@ -308,6 +308,20 @@ sealed abstract class IO[+E, +A] { self =>
   final def ensuring(finalizer: IO[Nothing, Unit]): IO[E, A] =
     new IO.Ensuring(self, finalizer)
 
+  /**
+   * Executes the action on the specified `ExecutionContext` and then shifts back
+   * to the default one.
+   */
+  final def on(ec: ExecutionContext): IO[E, A] =
+    IO.shift(ec) *> this <* IO.shift
```
Comment (RaasAhsan) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: I'm not as familiar with ZIO's error model yet, but could this an operation we may want to protect with `bracket`?

File: core/shared/src/main/scala/scalaz/zio/IO.scala (hunk 1 above)
Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: In Javascript, it _should_ mean `setImmediate`

File: core/jvm/src/main/scala/scalaz/zio/App.scala [hunk 5]
```diff
@@ -27,12 +25,7 @@ import scala.concurrent.duration.Duration
  */
 trait App extends RTS {
 
-  sealed trait ExitStatus
-  object ExitStatus {
-    case class ExitNow(code: Int)                         extends ExitStatus
-    case class ExitWhenDone(code: Int, timeout: Duration) extends ExitStatus
-    case object DoNotExit                                 extends ExitStatus
-  }
+  case class ExitStatus(code: Int)
```
Comment (jdegoes) [inline core/jvm/src/main/scala/scalaz/zio/App.scala]: Let's separate out this change into a separate PR, since I think there's some thought that

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 6]
```diff
@@ -308,6 +308,20 @@ sealed abstract class IO[+E, +A] { self =>
...
+   * to the default one.
+   */
+  final def on(ec: ExecutionContext): IO[E, A] =
+    IO.shift(ec) *> self <* IO.shift
+
+  /**
+   * Forks an action that will be executed on the specified `ExecutionContext`
+   * and then shifts back to the default one.
+   */
+  final def forkOn(ec: ExecutionContext): IO[E, Fiber[E, A]] =
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: I'm not sure this comment makes sense—because it's really only forking the `IO` _on_ the specified execution context. The main fiber which called `fork` never shifts off, so never needs to "shift back".

**Comment by RaasAhsan** [comment]
should be good to merge now

**Comment by jdegoes** [comment]
@RaasAhsan Looks like a few conflicts.

**Comment by RaasAhsan** [comment]
@jdegoes fixed, seems like it got a conflict from one of the other recent merges. thanks!

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 7]
```diff
@@ -334,6 +334,19 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
   final def ensuring(finalizer: IO[Nothing, Unit]): IO[E, A] =
     new IO.Ensuring(self, finalizer)
 
+  /**
+   * Executes the action on the specified `ExecutionContext` and then shifts back
+   * to the default one.
+   */
+  final def on(ec: ExecutionContext): IO[E, A] =
+    IO.shift(ec) *> self <* IO.shift
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: Oops, I missed this. It should definitely be done using `bracket` (if you want it to definitely return).

Also this version won't work. Exactly. It will never hop back. You need to create an `async` boundary that waits for the (forked) shifted IO to complete, then when it resumes, it will resume on its `rts`, thus returning to the original thread pool.

Comment (kaishh) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: shift submits a new task to the RTS ec, so, in my understanding, it should return to the RTS after `self`. (Also, finalizers currently run on separate fibers, shifts there wouldn't impact the parent AFAIK)

Comment (RaasAhsan) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]:
```
val ec = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())

def printThread: IO[Nothing, Unit] =
  IO.sync(println(Thread.currentThread()))

new RTS {}.unsafeRun {
  for {
    _ <- printThread
    _ <- printThread.on(ec)
    _ <- printThread
  } yield ()
}
```

Output:
```
Thread[main,5,main]
Thread[pool-1-thread-1,5,main]
Thread[zio-1-1468177767,5,zio]
```

> Also, finalizers currently run on separate fibers, shifts there wouldn't impact the parent AFAIK

Is this true? This definitely impacts `FiberLocal.locally` on cancellation. How come it forks for each finalizer?

Comment (kaishh) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: `.run` is used to collect defects in finalizer to propagate them to `ExitResult`, #313 might address the efficiency

Comment (RaasAhsan) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: Great, I think that will definitely help out, thanks!

On the other hand, there's an impressive set of things that will non-transparently fork and break `FiberLocal`, like `bracket0`. I will go through and create a list of them.

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 8]
```diff
@@ -339,7 +339,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * to the default one.
    */
   final def on(ec: ExecutionContext): IO[E, A] =
-    IO.shift(ec) *> self <* IO.shift
+    IO.shift(ec).bracket_(IO.shift)(self)
```
Comment (kaishh) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: It works, but in event of failure, it only does as a side-effect of .observe in `catchError`, when async is not completed by the time observe runs it will spawn a new task, but if it is completed the fiber will resume on the same thread. So this also shift back to the RTS even though there are no explicit async boundaries:

```scala
val ec = ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())

def printThread: IO[Nothing, Unit] =
  IO.sync(println(Thread.currentThread()))

new scalaz.zio.RTS {}.unsafeRun {
  for {
    _ <- printThread
    _ <- IO.shift(ec)
    _ <- printThread
    _ <- IO.unit.bracket_(IO.sync(Thread.sleep(1000)))(IO.fail(())).catchAll(_ => IO.unit)
    _ <- printThread
  } yield ()
}
```
```
Thread[main,5,main]
Thread[pool-1-thread-1,5,main]
Thread[zio-2-1169146729,5,zio]
```

File: core/shared/src/main/scala/scalaz/zio/IO.scala (hunk 7 above)
Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: That would be helpful. Then I think we have to eliminate as many of them as possible, or possibly, extend the `FiberLocal` model to handle forking, e.g. you automatically share the `FiberLocal` values of your parent fiber, in a copy-on-write model that ensures children only mutate `FiberLocal` locally.
